### PR TITLE
feat(`forge build`): add initcode size check

### PR DESCRIPTION
**Pull request by mgiagante**
Closes https://github.com/foundry-rs/foundry/issues/4783
described in the Additional Context of [this comment](https://github.com/foundry-rs/foundry/issues/4783#issue-1676759564)

## Motivation
Prior to [EIP-3860](https://eips.ethereum.org/EIPS/eip-3860) there was no size limit on creation code (initcode), and just a 24576 byte size limit on runtime code.
EIP-3860 went live in Shangai so now there is both:
- a 24576 byte size limit on runtime code
- a new 24576 * 2 = 49152 byte size limit on initcode.

Currently, `forge build --sizes` does not account for this new size limit.

## Solution
1. This PR adds a new column pair to the table displayed by `forge build --sizes` for the contract init code size (bytes) and the margin left counting towards the limit (bytes).
2. Also, the existing check for the deployed contract size is complemented with a new check for the init code size, making the process exist with status code 1 if the check fails.
3. Finally, a new option `--ignore-eip-3860` allows to ignore the check mentioned in the previous point. This is aimed to users deploying to chains that do not enforce this limit.

## Testing
- Two tests were adjusted to account for the new structure of the table.
- Two new tests were introduced to verify both failing and ignoring the new init code size limit.

### CODE REVIEW

**Comment by NelsonGaldeman** [comment]
Finally! Thank you ser 🫡

Tested and worked to me!

File: crates/common/src/compile.rs
Comment (grandizzy) [inline crates/common/src/compile.rs:400]: Since `initcode_size` is similar with `deployed_contract_size` let's refactor that one to `contract_size` and use for both runtime / init sizes, like

```diff
-/// Returns the size of the deployed contract
-pub fn deployed_contract_size<T: Artifact>(artifact: &T) -> Option<usize> {
-    let bytecode = artifact.get_deployed_bytecode_object()?;
+/// Returns the deployed or init size of contract.
+fn contract_size<T: Artifact>(artifact: &T, initcode: bool) -> Option<usize> {
+    let bytecode = if initcode {
+        artifact.get_bytecode_object()?
+    } else {
+        artifact.get_deployed_bytecode_object()?
+    };
     let size = match bytecode.as_ref() {

```

File: crates/forge/bin/cmd/build.rs
Comment (grandizzy) [inline crates/forge/bin/cmd/build.rs:66]: I think an alias as `--ignore-initcode-size` would be nice to have too

**Review by grandizzy — CHANGES_REQUESTED** [review]
Thank you! looks good, have added minor nits

File: crates/common/src/compile.rs
Comment (mgiagante) [reply crates/common/src/compile.rs]: Sure thing :+1: re-added.

**Review by grandizzy — APPROVED** [review]
thank you, lgtm! Pending other review

File: crates/common/src/compile.rs
Comment (zerosnacks) [inline crates/common/src/compile.rs:336]:
```suggestion
        // Filters out dev contracts (Test or Script)
```

**Review by zerosnacks — APPROVED** [review]
Thanks! Looks good 👍

File: crates/common/src/compile.rs
Comment (DaniPopes) [inline crates/common/src/compile.rs:48]: Can just be `bool`

File: crates/common/src/compile.rs
Comment (DaniPopes) [inline crates/common/src/compile.rs:354]:
```suggestion
                ..36_000 => Color::Reset,
                ..=CONTRACT_INITCODE_SIZE_LIMIT => Color::Yellow,
```

File: crates/forge/bin/cmd/build.rs
Comment (DaniPopes) [inline crates/forge/bin/cmd/build.rs:63]:
```suggestion
    /// Ignore initcode contract bytecode size limit introduced by EIP-3860.
```

File: crates/forge/tests/cli/test_helpers.rs
Comment (DaniPopes) [inline crates/forge/tests/cli/test_helpers.rs:1]: There is already a `utils` module, please move there

File: crates/forge/tests/cli/build.rs
Comment (DaniPopes) [inline crates/forge/tests/cli/build.rs:55]: Please also add assertions for stdout with `.stdout_eq(str![[r#""#]])` and then running the tests with `SNAPSHOT=overwrite` env

File: crates/forge/tests/cli/build.rs
Comment (DaniPopes) [inline crates/forge/tests/cli/build.rs:2]:
```suggestion
```

File: crates/common/src/compile.rs
Comment (DaniPopes) [inline crates/common/src/compile.rs:347]:
```suggestion
                ..18_000 => Color::Reset,
```

File: crates/common/src/compile.rs
Comment (DaniPopes) [inline crates/common/src/compile.rs:348]:
```suggestion
                ..=CONTRACT_RUNTIME_SIZE_LIMIT => Color::Yellow,
```

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/common/src/compile.rs
Comment (mds1) [inline crates/common/src/compile.rs:384]: The initcode check should probably also include the length of constructor arguments, since those become part of the initcode. But for dynamic length types we will have to make an assumption on length. One option is assume zero length dynamic types and let the margin column implicitly document how long dynamic type args can be. Another option is to continue excluding constructor arg length from the initcode length and document this limitation

File: crates/common/src/compile.rs
Comment (mgiagante) [reply crates/common/src/compile.rs]: Addressed. Thank you!

File: crates/forge/tests/cli/test_helpers.rs
Comment (mgiagante) [reply crates/forge/tests/cli/test_helpers.rs]: Addressed :heavy_check_mark:

File: crates/common/src/compile.rs
Comment (mgiagante) [reply crates/common/src/compile.rs]: Addressed :heavy_check_mark:

File: crates/forge/bin/cmd/build.rs
Comment (mgiagante) [reply crates/forge/bin/cmd/build.rs]: Done :heavy_check_mark:

File: crates/forge/tests/cli/build.rs
Comment (mgiagante) [reply crates/forge/tests/cli/build.rs]: I just pushed a commit adding the assertions.
The output we have right now displays a negative number telling by how many bytes the contract exceeds the size limit. But let me know if an error message instead of the table makes more sense.
Thank you!

File: crates/common/src/compile.rs
Comment (mgiagante) [reply crates/common/src/compile.rs:384]: I'm thinking we can continue excluding constructor arg length and document it in the `--help` output for `--sizes` like
```
    /// Print compiled contract sizes.
   /// Constructor argument length is not included in the calculation of initcode size.
    #[arg(long)]
    #[serde(skip)]
    pub sizes: bool,
```
does this sound good? @mds1

Comment (mgiagante) [reply crates/common/src/compile.rs:384]: @zerosnacks does the above look good to add it in this PR?

Comment (mds1) [reply crates/common/src/compile.rs:384]: > I'm thinking we can continue excluding constructor arg length and document it in the `--help` output for `--sizes` 

This is ok with me as long as @zerosnacks agrees :)

**Comment by mgiagante** [comment]
@DaniPopes Please let me know if this needs any further improvements before it can be merged. Thank you!

**Review by DaniPopes — APPROVED** [review]
Thanks!

**Pull request merged**
